### PR TITLE
[develop] Fix inconsistency with `pkg.list_pkgs` when using `attr` on RHEL systems

### PR DESCRIPTION
**Pull request by meaksh**
### What does this PR do?
As discussed, this PR brings https://github.com/saltstack/salt/pull/50563 to `develop` branch.

Now, `pkg_resource.format_pkg_list` does not directly parse the "arch" from the package name but delegates it to an optional `parse_arch_from_name` particular implementation on each virtual "pkg" module.

### CODE REVIEW

**Comment by thatch45** [comment]
At first glance it looks good to me, but @terminalmage should be the one to sign off on it.
I updated the branch, hopefully we can get a cleaner test run

File: salt/modules/pkg_resource.py [hunk 1]
```diff
@@ -311,21 +311,30 @@ def format_pkg_list(packages, versions_as_list, attr):
...
     if attr:
+        ret_attr = {}
         requested_attr = {'epoch', 'version', 'release', 'arch', 'install_date', 'install_date_time_t'}
 
         if attr != 'all':
-            requested_attr &= set(attr + ['version'])
+            requested_attr &= set(attr + ['version'] + ['arch'])
 
         for name in ret:
+            _parse_arch_from_name = __salt__.get('pkg.parse_arch_from_name', lambda pkgname: {'name': pkgname, 'arch': None})
```
Comment (terminalmage) [inline salt/modules/pkg_resource.py]: Let's not try to get too cute, here. What about something like this?

```python
if 'pkg.parse_arch_from_name' in __salt__:
    _parse_arch_from_name = __salt__['pkg.parse_arch_from_name'](pkgname)
else:
    _parse_arch_from_name = {'name': pkgname, 'arch': None}
```

Also, I have a _strong_ dislike of the function name `parse_arch_from_name`. We should think of something better.

**Review by terminalmage — CHANGES_REQUESTED** [review]
This doesn't even come close to satisfying the requirements outlined in https://github.com/saltstack/salt/pull/50563#issuecomment-442567698

File: salt/modules/pkg_resource.py (hunk 1 above)
Comment (dwoz) [reply salt/modules/pkg_resource.py]: I agree with @terminalmage here on both points. We should not be using lambdas for something like this. As for the method name, a simpler `parse_arch` seems fine to me.

Comment (meaksh) [reply salt/modules/pkg_resource.py]: Hey @dwoz , your feedback is now addressed. No more lambda and function has been renamed to `parse_arch`.

**Comment by meaksh** [comment]
Closing in favor of https://github.com/saltstack/salt/pull/55580

**Pull request closed without merging**